view/log.go

```diff
@@ -149,6 +149,136 @@ func (li *logView) Read(facts origins.Facts) (int, error) {
 	return l, nil
 }
 
+// merger maintains state of a merged stream of facts from multiple domains.
+type merger struct {
+	iterators []origins.Iterator
+	next      []*origins.Fact
+}
+
+func (m *merger) Next() (*origins.Fact, error) {
+	var maxTrn uint64 = 0
+	var maxI int = -1
+
+	// look at the next fact coming out of each of the the available logs,
+	// and choose the fact with the greatest transaction ID.
+	for i := 0; i < len(m.iterators); i++ {
+		if m.next[i] != nil && m.next[i].Transaction > maxTrn {
+			maxTrn = m.next[i].Transaction
+			maxI = i
+		}
+	}
+
+	if maxI == -1 {
+		return nil, io.EOF
+	}
+
+	// push the newest fact out of the merged stream
+	nextFact := m.next[maxI]
+	fact, err := m.iterators[maxI].Next()
+
+	if err != nil {
+		if err == io.EOF {
+			m.next[maxI] = nil
+		} else {
+			return nil, err
+		}
+	} else {
+		m.next[maxI] = fact
+	}
+
+	return nextFact, nil
+}
+
+func (m *merger) init() error {
+	var err error
+
+	for i := 0; i < len(m.iterators); i++ {
+		m.next[i], err = m.iterators[i].Next()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// merge multiple fact streams into one.
+// Note: iterators return facts in the order opposite to that in which
+// they were committed (i.e. newest facts with largest transaction IDs first),
+// and the Merge operation will preserve that.
+func Merge(iterators []origins.Iterator) (*merger, error) {
```

Comment:
I tried modifying Log.View/Now/Asof/Since methods to return origins.Iterator instead of *logView, and I only then realized that *logView implements not only the Iterator interface, but the Reader as well. 

Some of the callers to Log.View/Now/Asof/Since methods expect a Reader instance back, and they feed it to either ReadAll(r Reader) or ReadWriter(r Reader, w Writer). 

My first thought was that maybe the Reader and the Iterator can be combined into one interface, since they are both used for reading from a stream of facts. But I'll take another look, I'm quite possibly missing something. The fact that there is a matching Writer interface to go with the Reader probably complicates things.